### ISSUE TITLE
Bug: Command substitution overwrites/overloads machine env

### ISSUE THREAD

**Issue by hambergerpls** [dotenvx/dotenvx#268]
```sh
# .env
WORLD=$(echo hi)
HELLO=${WORLD}

WORLD=world dotenvx run -f .env -- env
# WORLD=world
# HELLO=hi

# Expected behavior
# WORLD=world
# HELLO=world
```
According to this [doc](https://dotenvx.com/docs/advanced/run-overload), adding `--overload` flag will overwrite machine env, but when used with command substitution without the `--overload` flag, it overwrites the machine env.

When `WORLD` is omitted from `.env` file, it gets the expected behavior.

Update:

I have to use `HELLO="$(echo ${WORLD})"` to get the expected behavior. I find this rather counter intuitive but understood that it runs as a subshell and uses the correct `WORLD`. 

I think a more intuitive approach is to prioritise the machine's environment variable and ignore the ones set in `.env` if exists, unless the `--overload` flag is passed which correctly describes the behavior: last environment variable seen wins.

**Comment by motdotla** [comment]
there might be a good reason for this as there is currently a test stating it should be there, but I need to dig in. could be historical from `dotenvx-expand`. i tend to agree initially with you though - seems counterintuitive. taking a look.

**Comment by hambergerpls** [comment]
> there might be a good reason for this as there is currently a test stating it should be behave as it currently does.
> 
> I need to dig in. could be historical from `dotenvx-expand`. i tend to agree initially with you though - seems counterintuitive.
https://github.com/dotenvx/dotenvx/blob/6e2c1f07c4f711425f3177200e07801e03b5053c/src/lib/helpers/parseDecryptEvalExpand.js#L25-L28

I played around with this by swapping the order of spreading the envs and it gets the expected behavior. However, it broke a lot quite a lot of tests.

Perhaps there should be an extra option like `--preserve` to preserve the machine env variables regardless if `--overload` is passed unless omitted. This change should not regress tests and previous issues.

**Comment by hambergerpls** [comment]
![image](https://github.com/dotenvx/dotenvx/assets/11707368/5f7e6243-572c-4ce0-bd3d-e56012d3259b)

@motdotla 

I added a `--preserve` option to preserve the machine env variables.

Tests are 100%, but I haven't wrote any tests for this specific feature.

I opened PR #269 for this change.

Let me know your comments!

**Comment by hambergerpls** [comment]
You're right, I'm also hesitant and dislike having to pass an extra option. I expect that preserving the machine env variables should be a default behavior. But it seems to be a major breaking change now.

I don't think dotenv-expand is the problem. Because it is doing exactly what it is supposed to do. It's just that `parseDecryptEvalExpand` treats the expansion as overload by default.

For instance, the intuitive priority for env variables is
```
Machine env > File env
```

When overloaded it's 
```
File env > Machine env
```

Having the `--preserve` flag would enforce this behavior without breaking anything. Unless you're okay with the breaking change, we can just treat the expansion without overload by default and not having to add an extra flag.

Or just stick with `HELLO=$(echo ${WORLD})` which is just not pleasant.

**Comment by motdotla** [comment]
yes, I see what you mean, but i think i didn't articulate the problem clearly enough with `dotenv-expand`:

### Currently

1. it expects to be the last function run after `process.env` has already been injected into
2. it doesn't have an `--overload` mechanism

For 1, this causes problems because `dotenvx` separates the concept of injection until the very last step - after all values have been parsed, substituted, and expanded. It does this for logging and `--overload` reasons.

For 2, because dotenv-expand doesn't have a concept of `--overload`, it has no way to properly respect if it should expand using the variable found in the file or already in process.env.

### Solution

I opted to rewrite dotenv-expand to track `parsed` and `processEnv` separately.

* `parsed` expands values using both `process.env` and the `file`
* if it finds the expansion variable in the `file` it treats that as the 'winner'
* otherwise, it uses the value from `process.env`

* `processEnv` expands values using both `process.env` and the `file`
* if it finds the expansion variable in `process.env` it treats that as the 'winner'
* otherwise, it falls back to using the expansion value in the `file`

https://github.com/dotenvx/dotenvx/pull/271/commits/b78359becb3bad6ffe25f3142f4f459a8f8bec6e#diff-e33f759627417bf8f44f0be1bfd7d0c640f1c8c6413cc2e5c5be52bc86882685R65

**Comment by motdotla** [comment]
fixed in `1.0.1` https://github.com/dotenvx/dotenvx/actions/runs/9704131361

```
# .env
WORLD=$(echo hi)
HELLO=${WORLD}

$ WORLD=world dotenvx get
{"WORLD":"world","HELLO":"world"}
```

**Comment by hambergerpls** [comment]
Can confirm that it is fixed. Thanks!
